### ISSUE TITLE
Need to allow "-" as a valid character for the text search

### ISSUE THREAD

**Issue by cmathieu333** [patrickkunka/mixitup#543]
I have a client using mixitup to filter their products (200 or so) and it works great. However when they include a "-" in the text search no results are shown. Normally this would be fine but many of their customers are repeat customers and actually know the product names including the hyphens. Is there anyway to allow the hyphen to be supported in a text search?

Here's a demo of it: https://topsoe-2115834.hs-sites.com/products/catalysts

Ex: "ar401" works but "ar-401" returns no results.

Thanks,
-Chris

**Comment by patrickkunka** [comment]
Hi Chris.

All special chars and punctuation are intentionally stripped out of search inputs to make the search less brittle.

The best way around the issue you're describing is simply to implement the same logic on your back-end to make sure that the class names or data attributes you are querying in your HTML also contain no special chars.

"ar-401" would therefore become "ar401". Its item in the DOM would look something like this:

```html
<div class="mix ar401">AR-401</div>
...
```
So, if I then enter "ar-401" in the search input, MixItUp will strip the "-" and query the DOM for "ar401", which will then match the "normalised" classname in your markup.

Hope that makes sense.

Some more info here: https://github.com/patrickkunka/mixitup/issues/402 (a closely-related issue)

**Comment by cmathieu333** [comment]
Hi Patrick,

Thanks for the response. So yes I can remove the hyphens from the class
names of the individual mix it up cards easy enough... The issue that the
end users on this site are going to include the hyphens in their searches
sometimes and this yields no product results. Is there a hack where I can
allow a hyphen? Or some other workaround?

https://topsoe-2115834.hs-sites.com/products/catalysts

Example a search for "ck" yields a bunch of results. A search for "ck-" or
"ck-302" (the actual product name) yields nothing.

-Chris


Christopher G. Mathieu
Chief Design Officer
<https://www.newbreedmarketing.com/>
T:802.655.0800 *|* M:802.236.4903
www.newbreedmarketing.com
Twitter <https://twitter.com/christophervt> | LinkedIn
<https://www.linkedin.com/in/cgmathieu/>

**Comment by patrickkunka** [comment]
Hi,

I think i assumed you were using the MultiFilter extension in my previous answer but maybe you're just using MixItUp core?

If so, the functionality I was referring to where all special characters are stripped out and queries are "normalised", is only included in the MultiFilter extension.

If you want to use the extension (which has built in text input support), it should solve the issue as all special chars (including hyphens) are removed from the end user queries, so "ck" and "ck-" should be exactly equivalent, and "ck-302" will be queried as "ck 302" - so that's why I was suggesting the approach of mirroring the same normalisation on the server side.

I hope that makes sense.

If you're not able to use the extension, I would recommend implementing a simple regular expression to replace all "non-word" characters with spaces as part of your keyup handler code (if you're copying the text input demo), e.g.:

```js
value = value.replace(/\W+/g, ' ');
```

But in both cases, replacing special chars on both the client and server side is the best way to ensure a robust and flexible search.

**Comment by jeffceriello** [comment]
Hi Patrick,

I am using the MultiFilter extension on my project and I am also trying to include the "-" in the text search. I have lowercased and removed the dash from the searchable string but when I search for any word with a dash in it I get no results.

This is how I have removed the dash from the data-attribute

```
<?php $title = get_the_title(); $lowercaseTitle = strtolower($title); $noDash = str_replace("-", "", $lowercaseTitle); ?>
<li <?php post_class( 'col-md-6 col-xl-4 mb-5 mix resource__item' ); ?> data-name="<?php echo $noDash; ?>">
```

and the text search field looks like this

```
<fieldset data-filter-group>
    <input type="text" data-search-attribute="data-name" placeholder="What are you looking for?"/>
</fieldset>
```

Here is a live link https://talkingfutures.clientapproval2.co.uk/resources/

If you type "levels", the first result would be T-levels. The data-name for that item is "tlevels" but if I search for t-levels I don't get any results.

Is there anything wrong I'm doing?

Thanks,
Jeff
